### PR TITLE
Release notes for 1.16

### PR DESCRIPTION
**Pull request by brson**
[Rendered](https://github.com/brson/rust/blob/relnotes/RELEASES.md)

### CODE REVIEW

File: RELEASES.md [hunk 1]
```diff
@@ -1,3 +1,182 @@
...
+[37057]: https://github.com/rust-lang/rust/pull/37057
+[37761]: https://github.com/rust-lang/rust/pull/37761
+[38006]: https://github.com/rust-lang/rust/pull/38006
+[38051]: https://github.com/rust-lang/rust/pull/38051
+[38062]: https://github.com/rust-lang/rust/pull/38062
+[38062]: https://github.com/rust-lang/rust/pull/38622
+[38066]: https://github.com/rust-lang/rust/pull/38066
+[38069]: https://github.com/rust-lang/rust/pull/38069
+[38131]: https://github.com/rust-lang/rust/pull/38131
+[38274]: https://github.com/rust-lang/rust/pull/
```
Comment (sanxiyn) [inline RELEASES.md]: Mistake here?

**Review by sanxiyn — CHANGES_REQUESTED** [review]

File: RELEASES.md [hunk 2]
```diff
@@ -1,3 +1,184 @@
...
+* [`File::set_permissions`]
+* [`String::split_off`]
+
+Libraries
+---------
+
+* [`[T]::binary_search` and `[T]::binary_search_by_key` now take
+  their argument by `Borrow` parameter][37761]
+* [All public types in std implement `Debug`][38006]
+* [`IpAddr` implements `From<Ipv4Adder>` and `From<Ipv6Addr>`][38327]
```
Comment (seppo0010) [inline RELEASES.md]: `Ipv4Addr`?

File: RELEASES.md [hunk 3]
```diff
@@ -1,3 +1,184 @@
+Version 1.16.0 (2017-03-16)
+===========================
+
+Language
+--------
+
+* Lifetimes in statics ad consts default to `'static`. [RFC 1623]
```
Comment (petrochenkov) [inline RELEASES.md]: typo ad -> and

File: RELEASES.md [hunk 4]
```diff
@@ -1,3 +1,184 @@
...
+--------
+
+* [`rustc` now supports `--emit=metadata`, which causes rustc to emit
+  a `.rmeta` file containing only crate metadata][38571]. This can be
+  used by tools like the Rust Language Service to perform
+  metadata-only builds.
+* [Fix `transmute::<T, U>` where `T` requires a bigger alignment than
+  `U`][38670]
+* [rustc: use -Xlinker when specifying an rpath with ',' in it][38798]
+* [Candidates for type suggestions draw from more sources][38927]
```
Comment (petrochenkov) [inline RELEASES.md]: This is an underestimation :)
It's not only type suggestions, Levenshtein based typo suggestions now work in most places, while previously they worked only for fields and *sometimes* for local variables. Together the overhaul of "no resolution"/"unexpected resolution" errors (https://github.com/rust-lang/rust/pull/38154) they result in large and systematic improvement in resolution diagnostics.

File: RELEASES.md [hunk 5]
```diff
@@ -1,3 +1,184 @@
...
+  specified in Cargo.toml][cargo/3546]
+* [crates.io will display categories listen in Cargo.toml][cargo/3301]
+* [Compilation profiles accept integer values for `debug`, in addition
+  to `true` and `false`. These are passed to `rustc` as the value to
+  `-C debuginfo`][cargo/3534]
+* [Implement `cargo --version --verbose`][cargo/3604]
+* [`Add `--dep-info` flag to output build dependencies compatible with
+  make and ninja][cargo/3557]
+* [Build all workspace members with `build --all`][cargo/3511]
+* [Document all workspace members with `document --all`][cargo/3515]
```
Comment (KalitaAlexey) [inline RELEASES.md]: It should be `doc --all`

File: RELEASES.md [hunk 6]
```diff
@@ -1,3 +1,184 @@
...
+-----
+
+* [crates.io will display CI badges from Travis and AppVeyor, if
+  specified in Cargo.toml][cargo/3546]
+* [crates.io will display categories listen in Cargo.toml][cargo/3301]
+* [Compilation profiles accept integer values for `debug`, in addition
+  to `true` and `false`. These are passed to `rustc` as the value to
+  `-C debuginfo`][cargo/3534]
+* [Implement `cargo --version --verbose`][cargo/3604]
+* [`Add `--dep-info` flag to output build dependencies compatible with
```
Comment (alexcrichton) [inline RELEASES.md]: Stray leading backtick here I believe

Comment (alexcrichton) [inline RELEASES.md]: Oh actually and this isn't a flag to cargo (despite the PR title), it happens automatically for all compilations now.

**Comment by alexcrichton** [comment]
r=me with typos and such fixed

File: RELEASES.md [hunk 7]
```diff
@@ -1,3 +1,184 @@
...
+  `Display`][38909]
+* [`Duration` implements `Sum`][38712]
+* [`String` implements `ToSocketAddrs`][39048]
+
+Cargo
+-----
+
+* [crates.io will display CI badges from Travis and AppVeyor, if
+  specified in Cargo.toml][cargo/3546]
+* [crates.io will display categories listen in Cargo.toml][cargo/3301]
```
Comment (wesleywiser) [inline RELEASES.md]: listen -> listed in?

File: RELEASES.md (hunk 4 above)
Comment (brson) [inline RELEASES.md]: Oh, oops I meant to say __"typo"__ suggestions. But I'll use your text. Thanks.

File: RELEASES.md (hunk 7 above)
Comment (brson) [inline RELEASES.md]: lol wow. thanks for the review

**Comment by brson** [comment]
Updated.

File: RELEASES.md [hunk 8]
```diff
@@ -1,3 +1,189 @@
...
+* [`Self` may appear in `impl` headers][38920]
+* [`Self` may appear in struct expressions][39282]
+
+Compiler
+--------
+
+* [`rustc` now supports `--emit=metadata`, which causes rustc to emit
+  a `.rmeta` file containing only crate metadata][38571]. This can be
+  used by tools like the Rust Language Service to perform
+  metadata-only builds.
```
Comment (eddyb) [inline RELEASES.md]: Isn't the most notable user `cargo check`, which can use `--emit=metadata` to check several crates quickly?

File: RELEASES.md [hunk 9]
```diff
@@ -1,3 +1,189 @@
...
+* [Build all workspace members with `build --all`][cargo/3511]
+* [Document all workspace members with `doc --all`][cargo/3515]
+* [Path deps outside workspace are not members][cargo/3443]
+
+Misc
+----
+
+* [`rustdoc` has a `--sysroot` argument that, like `rustc`, specifies
+  the path to the Rust implementation][38589]
+* [The `armv7-linux-androideabi` target no lenger enables NEON
```
Comment (killercup) [inline RELEASES.md]: s/lenger/longer

File: RELEASES.md (hunk 8 above)
Comment (brson) [inline RELEASES.md]: hm, is this the first release with `cargo check`?

File: RELEASES.md [hunk 10]
```diff
@@ -1,3 +1,189 @@
...
+  `Display`][38909]
+* [`Duration` implements `Sum`][38712]
+* [`String` implements `ToSocketAddrs`][39048]
+
+Cargo
+-----
+
+* [crates.io will display CI badges from Travis and AppVeyor, if
+  specified in Cargo.toml][cargo/3546]
+* [crates.io will display categories listed in in Cargo.toml][cargo/3301]
```
Comment (brson) [inline RELEASES.md]: "in in"

File: RELEASES.md (hunk 8 above)
Comment (eddyb) [inline RELEASES.md]: cc @nrc

Comment (alexcrichton) [inline RELEASES.md]: Yeah, Rust 1.16.0 is the first release with `cargo check`

**Review by sanxiyn — APPROVED** [review]

**Comment by brson** [comment]
Still needs to be updated for feedback.

**Comment by ishitatsuyuki** [comment]
No performance improvements? :o

**Comment by brson** [comment]
@ishitatsuyuki Sadly I didn't see anything of note, and the benchmarking service is down, so there's no data to draw from.

**Comment by golddranks** [comment]
Did omitting `'static` in statics land, though? I got the impression from https://github.com/rust-lang/rust/issues/35897#issuecomment-281460966 that it wasn't backported in the end!

**Comment by Aaronepower** [comment]
Also this mentions `writeln!, like println!, now accepts a form with no string or formatting arguments, to just print a newline`, however this is currently available in latest stable.

**Comment by llogiq** [comment]
It wasn't.

**Pull request merged**